### PR TITLE
MONGOID-5319 Permit driver FLE configuration to be specified in Mongoid config file


### PR DESCRIPTION
**Pull request by p-mongo**
https://jira.mongodb.org/browse/MONGOID-5319

### CODE REVIEW

File: spec/config/mongoid_with_schema_map_uuid.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+              properties:
+                message:
+                  encrypt:
+                    keyId:
+                      - !ruby/object:BSON::Binary
+                        data: !binary |-
+                          R/AgNcxASFiiJWKXqWGo5w==
+                        type: :uuid
+                    bsonType: "string"
+                    algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
```
Comment (johnnyshields) [inline spec/config/mongoid_with_schema_map_uuid.yml:26]: Very interesting... its good to support schema maps, but we should also use the `encrypts` macro introduced in Rails 7 (or something like it.) An analogy here is supporting indexes, which we also have a macro for.

File: spec/config/mongoid_with_schema_map_uuid.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+              key: "z7iYiYKLuYymEWtk4kfny1ESBwwFdA58qMqff96A8ghiOcIK75lJGPUIocku8LOFjQuEgeIP4xlln3s7r93FV9J5sAE7zg8U"
+          schema_map:
+            blog_development.comments:
+              properties:
+                message:
+                  encrypt:
+                    keyId:
+                      - !ruby/object:BSON::Binary
+                        data: !binary |-
+                          R/AgNcxASFiiJWKXqWGo5w==
```
Comment (johnnyshields) [inline spec/config/mongoid_with_schema_map_uuid.yml:23]: Where is this generated from? Any reason we can't store UUIDs in a human-readable form in the config?

**Review by Neilshweky — APPROVED** [review]

File: docs/reference/configuration.txt [hunk 3]
```diff
@@ -725,6 +725,95 @@ be executed sequentially during socket creation.
...
+              blog_development.comments:
+                properties:
+                  message:
+                    encrypt:
+                      keyId:
+                        - !ruby/object:BSON::Binary
+                          data: !binary |-
+                            R/AgNcxASFiiJWKXqWGo5w==
+                          type: :uuid
+                      bsonType: "string"
```
Comment (johnnyshields) [inline docs/reference/configuration.txt:761]: `bsonType` should use `snake_case` (`bson_type`).

File: docs/reference/configuration.txt [hunk 4]
```diff
@@ -725,6 +725,95 @@ be executed sequentially during socket creation.
...
+            key_vault_namespace: 'keyvault.datakeys'
+            kms_providers:
+              local:
+                key: "z7iYiYKLuYymEWtk4kfny1ESBwwFdA58qMqff96A8ghiOcIK75lJGPUIocku8LOFjQuEgeIP4xlln3s7r93FV9J5sAE7zg8U"
+            schema_map:
+              blog_development.comments:
+                properties:
+                  message:
+                    encrypt:
+                      keyId:
```
Comment (johnnyshields) [inline docs/reference/configuration.txt:756]: Should use `snake_case`

File: docs/reference/configuration.txt [hunk 5]
```diff
@@ -725,6 +725,95 @@ be executed sequentially during socket creation.
...
+                        - !ruby/object:BSON::Binary
+                          data: !binary |-
+                            R/AgNcxASFiiJWKXqWGo5w==
+                          type: :uuid
+                      bsonType: "string"
+                      algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
+                bsonType: "object"
```
Comment (johnnyshields) [inline docs/reference/configuration.txt:763]: Should use snake_case

File: spec/support/schema_maps/schema_map_kmip.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+  "properties": {
+    "ssn": {
+      "encrypt": {
+        "keyId": [{
+          "$binary": {
+            "base64": "KMIPAAAAAAAAAAAAAAAAAA==",
+            "subType": "04"
```
Comment (johnnyshields) [inline spec/support/schema_maps/schema_map_kmip.json:8]: Should use `snake_case`

File: spec/config/mongoid_with_schema_map_uuid.yml (hunk 1 above)
Comment (p-mongo) [reply spec/config/mongoid_with_schema_map_uuid.yml:26]: Please feel free to add this comment to https://jira.mongodb.org/browse/MONGOID-5248 which covers an implementation of FLE in Mongoid.

File: spec/config/mongoid_with_schema_map_uuid.yml (hunk 2 above)
Comment (p-mongo) [reply spec/config/mongoid_with_schema_map_uuid.yml:23]: This particular value came from https://github.com/mongodb/mongo-ruby-driver/pull/2429. To learn how FLE works in driver, please refer to https://www.mongodb.com/docs/ruby-driver/master/reference/client-side-encryption/.

**Review by Neilshweky — APPROVED** [review]

**Review by comandeo — APPROVED** [review]

**Pull request merged**